README.md
# HUDstats Car Pooling API

The service should provide a JSON API that extends Icelandic car pooling API
http://docs.apis.is/#endpoint-rides which will show passengers and drivers along with the details about
each ride.


## Usage

Start the server by running the command below. 

The default port is 3000. You should be able to navigate to localhost:3000/drivers or localhost:3000/passengers on your browser.

The carpooling API will now be extended with additional fields

    $ lein run

## Testing

    $ bin/kaocha --watch





## License

The MIT License

Copyright (c) 2021 Tovieye Moses Ozi

Permission is hereby granted, free of charge, 
to any person obtaining a copy of this software and 
associated documentation files (the "Software"), to 
deal in the Software without restriction, including 
without limitation the rights to use, copy, modify, 
merge, publish, distribute, sublicense, and/or sell 
copies of the Software, and to permit persons to whom 
the Software is furnished to do so, 
subject to the following conditions:

The above copyright notice and this permission notice 
shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR 
ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-08-23
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-08-23
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/hudstats/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/hudstats/compare/0.1.0...0.1.1


test/hudstats/core_test.clj
(ns hudstats.core-test
  (:require [hudstats.core :as hud]
            [hudstats.mocks :as mock-data]
            [clj-http.fake :refer [with-fake-routes]]
            [jsonista.core :as j]
            [ring.mock.request :as mock]
            [clojure.test :refer [deftest are is]]))

(deftest process-page
  (is (= {"Name" "Keflaví­k"
          "Phone" "0124568"
          "Seats" "2"}
         (hud/process-page mock-data/page-tree))))

(deftest aggregate
  (is (= {"From" "Keflaví­k", "Requesting" "Passengers", "To" "Selfoss"} (reduce hud/seq->map {} mock-data/processed-page))))

(deftest server
  (with-fake-routes (merge mock-data/http-mocks {hud/drivers-url (fn [_req] {:status 200 :headers {} :body (j/write-value-as-string mock-data/driver-results)})
                                                 hud/passengers-url (fn [_req] {:status 200 :headers {} :body (j/write-value-as-string mock-data/passenger-results)})})

    (are [result method path] (= result (-> (mock/request method path) hud/app (update :body #(j/read-value % j/default-object-mapper))))

      {:status 200 :body  mock-data/driver-results :headers {"Content-Type" "application/json; charset=utf-8"}} :get "/drivers"
      {:status 200 :body  mock-data/passenger-results :headers {"Content-Type" "application/json; charset=utf-8"}} :get "/passengers")))



project.clj
(defproject hudstats "0.1.0-SNAPSHOT"
  :description "service should provide a JSON API that extends Icelandic car pooling API
http://docs.apis.is/#endpoint-rides which will show passengers and drivers along w"
  :license {:name "MIT"
            :url "https://mit-license.org/"}
  :dependencies [[org.clojure/clojure "1.10.3"]
                 [metosin/reitit "0.5.15"]
                 [ring/ring-jetty-adapter "1.9.4"]
                 [clj-http "3.12.3"]
                 [zerg000000/simple-cors "0.0.8"]
                 [factual/geo "3.0.1"]
                 [enlive "1.1.6"]]
  :main ^:skip-aot hudstats.core
  :target-path "target/%s"
  :aliases {"kaocha" ["run" "-m" "kaocha.runner"]}
  :profiles {:uberjar {:aot :all}
             :dev [:testing]
             :test [:testing]
             :testing {:dependencies [[ring/ring-mock "0.4.0"]
                                      [clj-http-fake "1.0.3"]
                                      [lambdaisland/kaocha "1.0.861"]]
                       :jvm-opts ["-Dsnoop.enabled"]}})


src/hudstats/distance.clj
(ns hudstats.distance
  (:require [geo.spatial :as spatial] )
  (:import java.text.DecimalFormat))

(def default-city [64.13548 -21.89541])
(def distance-format (DecimalFormat. "#"))

(def cities {"Reykjaví­k" [64.13548 -21.89541]
             "Akureyi" [65.68353 -18.0878]
             "Akranes" [64.32179 -22.0749]
             "Vestmannaeyjar" [63.44273 -20.27339]
             "Kopavogur" [64.11234 -21.91298]
             "Gardabaer" [64.08865 -21.92298]
             "Selfoss" [63.93311 -20.99712]
             "Saudarkrokur" [65.74611 -19.63944]
             "Grindaví­k" [63.83849 -22.43931]
             "Hafnarfjordur" [64.0671 -21.93774]
             "Mosfellsbaer" [64.16667 -21.7]
             "Seltjarnarnes" [64.15309 -21.99499]
             "Ísafjörður" [66.07475 -23.13498]
             "Keflavík" [63.9998 -22.5583]
             "Keflavík (Airport)" [63.985 -22.6282]
             "Hellissandur" [64.9163 -23.8757]
             "Egilsstaðir" [65.2669 -14.3948]
             "Höfn" [64.290665504 -15.222999108]
             "Hella" [63.835 -20.3919]
             "Húsaví­k" [65.4041 -13.6981]
             "Landmannalaugar" [63.9830 -19.0670]
             "Hvolsvöllur" [63.749997 -20.2333324]
             "Reykholt" [64.6657 -21.2871]
             "Mývatn" [65.6039 -16.9961]
             "Kirkjubæjarklaustur" [63.7833302 -18.0666664]
             "Vík" [63.4186 -19.0060]
             "Sandgerði" [64.051833126 -22.70249719]})

(defn distance [from to]
  (let [from-encoded (apply spatial/spatial4j-point (get cities from default-city))
        to-encoded (apply spatial/spatial4j-point (get cities to default-city))]
    (str (.format distance-format (/ (spatial/distance from-encoded to-encoded) 1000)) "km")))

src/hudstats/core.clj
(ns hudstats.core
  (:gen-class)
  (:require [reitit.ring :as ring]
            [muuntaja.core :as m]
            [simple-cors.ring.middleware :as cors]
            [net.cgrand.enlive-html :as html]
            [reitit.ring.coercion :as rrc]
            [clj-http.client :as http]
            [hudstats.distance :refer [distance]]
            [clojure.string :as str]
            [reitit.coercion.malli]
            [jsonista.core :as j]
            [reitit.ring.middleware.muuntaja :as muuntaja]
            [reitit.ring.middleware.exception :as exception]
            [ring.adapter.jetty :as jetty]))

(defn get-base-data [url]
  (some-> (http/get url)
          :body
          (j/read-value j/keyword-keys-object-mapper)))

(def drivers-url "http://apis.is/rides/samferda-drivers/")

(def passengers-url "http://apis.is/rides/samferda-passengers/")

(def extra-keys ["Seats" "Name" "Phone" "Mobile" "E-mail" "Non-smoke car"])

(defn fetch-url [url]
  (html/html-resource (java.net.URL. url)))

(defn extract-links [results] (map :link results))

(defn extract-key [v]
  (-> v :content first :content first (str/replace ":" "")))

(defn extract-val [v]
  (-> v :content first))

(defn seq->map [acc a]
  (assoc acc (extract-key (first a)) (extract-val (second a))))

(defn process-page [page-tree]
  (as-> page-tree $ (map  (comp #(filter map? %)  :content) $)
       (reduce seq->map {} $)
       (select-keys $ extra-keys)))

(defn enhance-api [url]
  (let [results (:results (get-base-data url))
        distances (map (fn [{:keys [from to]}]  (distance from to)) results)
        link-pages (doall (pmap (fn [{link :link}] (fetch-url link)) results))]
    (map-indexed (fn [index page] (->> (html/select page [:tr]) process-page (merge (nth results index) {:distance (nth distances index)}) )) link-pages)))

(defn make-handler [url]
  (fn [_]
    {:status 200
     :body  {:results (enhance-api url)}}))

(def drivers-handler (make-handler drivers-url))
(def passengers-handler (make-handler passengers-url))

(def app
  (ring/ring-handler
   (ring/router
    [["/drivers" {:get {:handler  drivers-handler}}]
     ["/passengers" {:get {:handler passengers-handler}}]]
    {:data {:coercion reitit.coercion.malli/coercion
            :muuntaja   m/instance
            :middleware [[cors/wrap {:cors-config {:origins "*"}}]
                         muuntaja/format-negotiate-middleware
                         muuntaja/format-request-middleware
                         muuntaja/format-response-middleware
                         exception/exception-middleware
                         rrc/coerce-response-middleware]}})
   (ring/create-default-handler)))

(defn start [{:keys [port] :or {port 3000}}]
  (let [server (jetty/run-jetty #'app {:port port, :join? false})]
    (println "server running on port" port)
    server))

(defn -main
  [& args]
  (start {}))

(comment
  (def server (start {}))
  (.stop server)
  (require '[ring.mock.request :as mock])
  (require '[jsonista.core :as j])
  (def api-response (-> (http/get drivers-url)
                        :body
                        (j/read-value j/keyword-keys-object-mapper)))


  "")


